### PR TITLE
Tensorboard callback modifications

### PR DESCRIPTION
**Pull request by sameermanek**
This PR adds the option to write batch-level performance to Tensorboard. It's associated w/ this [thread](https://github.com/fchollet/keras/issues/6692). Apologies if this should go elsewhere or requires more discussion before creating a PR. Thanks

### CODE REVIEW

**Comment by sameermanek** [comment]
- This could modify existing behavior:

1. It adds an argument `write_batch_performance` to the Tensorboard callback constructor (defaults to False). It's currently in the middle of the argument list, near the (logically) related arguments, but this could change the behavior if any users are relying on positional rather than named arguments here. I think this is unlikely because there were already 9 named arguments, all of which have default values (and the usage examples I can find all use named arguments). If it makes more sense, I can move this to the end of the argument list, which would eliminate the behavior change. 

2. This also slightly changes the x-axis on the Tensorboard summaries to be the number of examples seen instead of the number of epochs

- This is being incorporated into the existing Tensorboard tests (in test_callbacks.py). 

Let me know if I should change any of these. Thanks

**Comment by bdwyer2** [comment]
@sameermanek thanks for doing this! Have you verified that the validation statistics are correctly reported with this PR?

**Comment by sameermanek** [comment]
@bdwyer2 Sorry, I don't understand -- what are the validation statistics? I ran all the tests locally and they succeeded (so those matched the Travis CI build tests) if that's what you're referring to. Thanks.

**Comment by ViaFerrata** [comment]
Question: Did you investigate the performance impact of logging after each batch? For RAM it shouldn't matter (even with >10 Mio. dataset like I use), but what about computation time?
If it is significant, one could think about an option to e.g. only log every hundredth batch.

**Comment by karantan** [comment]
Will this allow using `histogram_freq=1` option in TensorBoard callback when using `fit_generator`?

**Comment by sameermanek** [comment]
@ViaFerrata I ran a quick (totally non-scientific) test, and the performance impact was nearly negligible. I trained a network with and without the batch-level logging (alternating between with and without batch-level callback, 5 iterations each, 98 batches and 100 epochs each iteration)

Without batch-level logging: 142.4 seconds per iteration on average, sd of 7.1 seconds
With batch-level logging: 143.9 seconds per iteration, sd of 5.7 seconds

@karantan I'm not sure I follow -- this doesn't address any outstanding issues if that's what you're referring to.

**Comment by cdancette** [comment]
This PR would be very useful, any idea when it will be merged ?

**Comment by chipgarner** [comment]
Please! I am running @sameermanek 's version, Keras needs this.

**Comment by rafaspadilha** [comment]
This PR would be very useful!

**Comment by chipgarner** [comment]
This is working well for us on Ubuntu 16.04, Windows 10, and AWS servers using their Deep Learning AMI (Ubuntu) and modifying the one file. Needs one approved review!

**Comment by ahundt** [comment]
I've tested this and it is working well.

File: keras/callbacks.py [hunk 1]
```diff
@@ -776,7 +781,7 @@ def on_epoch_end(self, epoch, logs=None):
                     feed_dict = dict(zip(tensors, batch_val))
                     result = self.sess.run([self.merged], feed_dict=feed_dict)
                     summary_str = result[0]
-                    self.writer.add_summary(summary_str, epoch)
+                    self.writer.add_summary(summary_str, self.seen)
```
Comment (ahundt) [inline keras/callbacks.py]: If write batch performance is false self.seen should probably be equal to the epoch so the current behavior remains unchanged.

File: keras/callbacks.py [hunk 2]
```diff
@@ -792,12 +797,31 @@ def on_epoch_end(self, epoch, logs=None):
...
+                    continue
+                summary = tf.Summary()
+                summary_value = summary.value.add()
+                summary_value.simple_value = value.item()
+                summary_value.tag = name
+                self.writer.add_summary(summary, self.seen)
+            self.writer.flush()
+
+        self.seen += self.batch_size
+    # @todo write some unit tests for this
```
Comment (ahundt) [inline keras/callbacks.py]: I believe you can remove the todo here, since you set it to true in the unit test

File: keras/callbacks.py [hunk 3]
```diff
@@ -792,12 +797,31 @@ def on_epoch_end(self, epoch, logs=None):
             summary_value = summary.value.add()
             summary_value.simple_value = value.item()
             summary_value.tag = name
-            self.writer.add_summary(summary, epoch)
+            self.writer.add_summary(summary, self.seen)
```
Comment (ahundt) [inline keras/callbacks.py]: this should also follow the epoch behavior if batch mode is disabled

File: keras/callbacks.py [hunk 4]
```diff
@@ -611,6 +611,8 @@ class TensorBoard(Callback):
             for histograms computation.
         write_images: whether to write model weights to visualize as
             image in TensorBoard.
+        write_batch_performance: whether to write training metrics on batch
```
Comment (ahundt) [inline keras/callbacks.py]: Can we change this parameter to be `write_step` with the options `'epoch'` and `'batch'`, and improve the description?

I think `write_step` might be more clear, and doesn't break the true/false setting in the future.

File: keras/callbacks.py [hunk 5]
```diff
@@ -792,12 +797,31 @@ def on_epoch_end(self, epoch, logs=None):
...
+        logs = logs or {}
+
+        if self.write_batch_performance:
+            for name, value in logs.items():
+                if name in ['batch', 'size']:
+                    continue
+                summary = tf.Summary()
+                summary_value = summary.value.add()
+                summary_value.simple_value = value.item()
+                summary_value.tag = name
```
Comment (ahundt) [inline keras/callbacks.py]: What about prepending `batch_` to the name?
Then separate batch and epoch metrics can both be written.

Actually, that might allow the write_batches parameter to be dropped entirely as the existing behavior wouldn't be affected. Which do you think is better, this or `write_step`?

File: keras/callbacks.py (hunk 4 above)
Comment (ahundt) [reply keras/callbacks.py]: Alternately, might it be better to write both and prepend `epoch_` or `batch_` accordingly?

File: keras/callbacks.py [hunk 6]
```diff
@@ -807,9 +807,9 @@ def on_train_end(self, _):
     def on_batch_end(self, batch, logs=None):
         logs = logs or {}
 
-        if self.write_batch_performance == True:
+        if self.write_batch_performance:
```
Comment (wenmin-wu) [inline keras/callbacks.py]: It's better use a flag to control write epoch-level summary or batch-level summary. The batch-level summary will be overwrite in your implementation.

**Comment by Faur** [comment]
Bump. 

I think this is very useful, and it looks like most of the work has been done. Would really like to see this implemented, thank you!

**Comment by nandaw** [comment]
This would be really helpful for us working with huge datasets. Is there anything that I can do to help?

**Comment by fchollet** [comment]
I still think this is a good feature to add. Sorry for the delay in processing the PR.

@omalleyt12 : this should be your area of expertise. Could you please take it over?

**Comment by omalleyt12** [comment]
@fchollet yep will do!

**Comment by gabrieldemarmiesse** [comment]
Closing this PR in favor of #11152. Feel free to reopen it or comment if necessary.

**Comment by gabrieldemarmiesse** [comment]
This feature is now implemented and merged in master.
